### ISSUE TITLE
Async methods missing for FakeContext?

### ISSUE THREAD

**Issue by slee-accesso** [ReversePOCO/EntityFramework-Reverse-POCO-Code-First-Generator#673]
I've been trying to use the FakeDbContext, and it fails anytime I try to run async methods. The synchronous methods work - and the async methods do work against the "real" context, just not against the fake one. 

Is there a setting I'm missing? Or is the generation code skipping the async methods? Here's the sample code I'm using in a unit test:

```csharp
var factory = new Project.Data.Context.FakeProjectContextFactory();
var context = factory.CreateDbContext();
var c1 = context.Contacts.FirstOrDefault();    // succeeds (get a null back)
var c2 = await context.Contacts.FirstOrDefaultAsync();  // Dies a horrible death. Exception below.
```
Exception:

```
System.ArgumentException
Argument expression is not valid
   at System.Linq.EnumerableQuery`1.System.Linq.IQueryProvider.Execute[TElement](Expression expression)
   at Project.Data.Context.FakeDbAsyncQueryProvider`1.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken) in C:\Dev\Project\Project.Data\Project.Data.Context\GeneratedCode\FakeDbSet.cs:line 239
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ExecuteAsync[TSource,TResult](MethodInfo operatorMethodInfo, IQueryable`1 source, Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ExecuteAsync[TSource,TResult](MethodInfo operatorMethodInfo, IQueryable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.FirstOrDefaultAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at UnitTests.ContextFactoryTests.FakeContextFactory_AddTwoContactsAsync_EnsureTheyWereAdded() in C:\Dev\Project\Project.Data\UnitTests\FactoryTests.cs:line 78
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_0(Object state)
   at Xunit.Sdk.AsyncTestSyncContext.<>c__DisplayClass7_0.<Post>b__1(Object _) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\AsyncTestSyncContext.cs:line 75
```
This occurs for every async method I've tried (AddRangeAsync, query.ToArrayAsync(), query.ToListAsync(), etc.)

And help would be much appreciated! Thanks!

  - Sean

**Comment by sjh37** [comment]
In the generated code, change one function within `public class FakeDbAsyncQueryProvider<TEntity> : IAsyncQueryProvider` to be:

```c#
public TResult ExecuteAsync<TResult>(Expression expression, CancellationToken cancellationToken = new CancellationToken())
{
    var expectedResultType = typeof(TResult).GetGenericArguments()[0];
    var executionResult = typeof(IQueryProvider)
        .GetMethod(
            name: nameof(IQueryProvider.Execute),
            genericParameterCount: 1,
            types: new[] { typeof(Expression) })
        .MakeGenericMethod(expectedResultType)
        .Invoke(this, new[] { expression });

    return (TResult) typeof(Task).GetMethod(nameof(Task.FromResult))
        ?.MakeGenericMethod(expectedResultType)
        .Invoke(null, new[] { executionResult });
}
```

Let me know if that works for you. In the mean time I will implement this in the generator.

**Comment by sjh37** [comment]
See changeset https://github.com/sjh37/EntityFramework-Reverse-POCO-Code-First-Generator/commit/f0a85890d2e37f3b63ab9601439daf33c0858c94

Grab the latest code for [EF.Reverse.POCO.v3.ttinclude](https://raw.githubusercontent.com/sjh37/EntityFramework-Reverse-POCO-Code-First-Generator/master/EntityFramework.Reverse.POCO.Generator/EF.Reverse.POCO.v3.ttinclude) and update your project.

**Comment by sjh37** [comment]
Let me know if this does not work, and I will re-open this case.

**Comment by slee-accesso** [comment]
Simon - thanks for such a quick response! 

Your fix took care of 3 of my test cases, but one still fails - AddRangeAsync(). FirstOrDefaultAsync(), ToArrayAsync() and ToListAsync() all now work as expected. Here's the code for the failing AddRangeAsync() test:

```csharp
[Fact]
public async void FakeContextFactory_AddFourContactsWithAddRangeAsync_EnsureTheyWereAdded()
{
    var list1 = new List<Contact>
    {
        new Contact { FirstName = "TestOne", LastName = "Test", ContactId = 1},
        new Contact { FirstName = "TestTwo", LastName = "Test", ContactId = 2}
    };
    var list2 = new List<Contact>
    {
        new Contact { FirstName = "TestThree", LastName = "Test", ContactId = 3},
        new Contact { FirstName = "TestFour", LastName = "Test", ContactId = 4}
    };
    var factory = new Project.Data.V5.Context.FakeProjectContextFactory();
    var context = factory.CreateDbContext();
    context.Contacts.AddRange(list1);            // This works fine
    await context.Contacts.AddRangeAsync(list2); // Crashes - exception below...
    Assert.Equal(4, context.Contacts.Count());
}
```

Exception:
```
System.NotImplementedException
The method or operation is not implemented.
   at Microsoft.EntityFrameworkCore.DbSet`1.AddRangeAsync(IEnumerable`1 entities, CancellationToken cancellationToken)
   at UnitTests.ContextFactoryTests.FakeContextFactory_AddFourContactsWithAddRangeAsync_EnsureTheyWereAdded() in C:\Dev\Project\Project.Data\UnitTests\ContextFactoryTests.cs:line 83
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_0(Object state)
   at Xunit.Sdk.AsyncTestSyncContext.<>c__DisplayClass7_0.<Post>b__1(Object _) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\AsyncTestSyncContext.cs:line 75
```

**Comment by sjh37** [comment]
@slee-accesso This is now fixed and will be in the next release.

**Comment by sjh37** [comment]
Released in v3.4.1
